### PR TITLE
Remove the fallback to the legacy coherency algorithm.

### PR DESCRIPTION
**Pull request by andriipatsula**
https://github.com/dotnet/arcade/issues/6782

### CODE REVIEW

**Comment by mmitche** [comment]
We need to implement an alert mechanism when this fails. Right now, Legacy coherency is probably filling some gaps where repos are failing the strict coherency update. Whether it gets it right is unknown, but they're getting updates rather than failing silently.

We should implement this part: https://github.com/dotnet/arcade/issues/6782#issuecomment-1426334037 in this PR:

- We allow coherency to fail and still do an update PR. This is not the case today.
- In the update PR, we add a new Maestro policy check that is "coherency valid" (sort of like the Do not automerge downgrades). This is always on, regardless of repo policies. If coherency is failing, then we make this check Red

This behavior would not be implemented in local darc, just the subscription actor service. The check, if it could not get info on which dependencies fail coherency, would tell the user to go and run local darc to resolve.

**Comment by andriipatsula** [comment]
I am going to introduce a new merge policy class like:
```C#
public class ValidateCoherencyMergePolicy : MergePolicy
{
    public override string DisplayName => "TODO";

    /// overrides `IMergePolicy.EvaluateAsync`
    public override async Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemote darc)
    {
        string prBranch = "";
        string repoUrl = "";

        List<DependencyDetail> existingDependencies = (await darc.GetDependenciesAsync(repoUrl, prBranch)).ToList();
        List<DependencyUpdate> coherencyUpdates = new List<DependencyUpdate>();
        try
        {
            // remoteFactory ???
            coherencyUpdates = await darc.GetRequiredCoherencyUpdatesAsync(existingDependencies, remoteFactory, CoherencyMode.Strict);
        }
        catch (DarcCoherencyException)
        {
            return Fail("Is not coherent");
        }

        if (coherencyUpdates.Any())
        {
            return Fail("Is not coherent");
        }

        return Succeed("Is coherent.");

    }
}
```

and utilize the [`IRemote.GetRequiredCoherencyUpdatesAsync(IEnumerable<DependencyDetail>, IRemoteFactory, CoherencyMode)`](https://github.com/dotnet/arcade-services/blob/main/src/Microsoft.DotNet.Darc/src/DarcLib/IRemote.cs#L293) to get the list of coherency update and if it empty - the repo + pr branch is coherent.


[!] But we don't have access to `IRemoteFactory` on the `MergePolicy` level. What about to modify the interface of the `IMergePolicy` interface?
```diff
public interface IMergePolicy : IMergePolicyInfo
{
-   Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemote darc);
+   Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemoteFactory remoteFactory);
}
```

**Comment by mmitche** [comment]
I don't think you need to redo the coherency calculation. Instead, GetRequiredUpdates would return whether the coherency update failed: https://github.com/dotnet/arcade-services/blob/main/src/Maestro/SubscriptionActorService/PullRequestActor.cs#L1083-L1088

Then, you modify the IPullRequest interface to include a bit that indicates whether the last coherency update failed or not. It currently just knows about the set of updates in the PR.

The new merge policy would check this bit and let the user know.

**Comment by andriipatsula** [comment]
@riarenas , @mmitche could you please short feedback about the latest changes in the PR. @riarenas , we can have a call and discuss next steps.

1. Example of a test PR with a failing coherency check: https://github.com/maestro-auth-test/maestro-test3/pull/2059/checks?check_run_id=13442834926
2. I am going to add one more test scenario to ScenarioTests_GitHubFlow.cs
3. As we are going to remove fallback and introduce new check, this new Github check won't be available in existing subscription (as I understand). Do I need to update all existing subscriptions? Write a data change script?

**Comment by riarenas** [comment]
> 1. Example of a test PR with a failing coherency check: https://github.com/maestro-auth-test/maestro-test3/pull/2059/checks?check_run_id=13442834926

Is it possible to add instructions to the check on how to fix the coherency issue? I believe the error gives some more details that would be helpful to add.  

> 2. I am going to add one more test scenario to ScenarioTests_GitHubFlow.cs

Sounds good
> 3. As we are going to remove fallback and introduce new check, this new Github check won't be available in existing subscription (as I understand). Do I need to update all existing subscriptions? Write a data change script?

We could add the check to the standard merge policy, which is already an aggregation of policies that we suggest everyone should have: https://github.com/dotnet/arcade-services/blob/main/src/Maestro/Maestro.MergePolicies/StandardMergePolicy.cs

For subscriptions not using the standard policy I see a few options:

* we just inform partners that this is a new check and to please add it to their subscriptions.
* We update every subscription (ideally through the API or through darc, not through the database) to include the new policy
* We inject the policy and check to every subscription, regardless of how they have them configured.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 1]
```diff
@@ -10,12 +10,12 @@
 using System.Runtime.Serialization;
 using System.Text;
 using System.Threading.Tasks;
+using Castle.Core.Internal;
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:13]: Where is this used?

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:13]: I was wondering the same and checked out the branch. Looks like it's for the IsNullOrEmpty Extension Methods used in LNS 748 and 973

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 2]
```diff
@@ -742,10 +742,10 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
             (string targetRepository, string targetBranch) = await GetTargetAsync();
             IRemote darcRemote = await DarcRemoteFactory.GetRemoteAsync(targetRepository, Logger);
 
-            List<(UpdateAssetsParameters update, List<DependencyUpdate> deps)> requiredUpdates =
+            TargetRepoDependencyUpdate repoDependencyUpdate =
                 await GetRequiredUpdates(updates, DarcRemoteFactory, targetRepository, targetBranch);
 
-            if (requiredUpdates.Count < 1)
+            if (repoDependencyUpdate.CoherencyCheckSuccessful && repoDependencyUpdate.RequiredUpdates.IsNullOrEmpty())
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:748]: This looks like we'd create a PR if the coherency check currently fails? If so, this is great. Also, do we have a test for this corner case?

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 1 above)
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:13]: Would it be better to write our own or use this one? We already use Castle.Core, but this still seems a little out of place.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 3]
```diff
@@ -755,13 +755,17 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
 
             try
             {
-                string description = await CalculatePRDescriptionAndCommitUpdatesAsync(requiredUpdates, null, DarcRemoteFactory, targetRepository, newBranchName);
+                string description = await CalculatePRDescriptionAndCommitUpdatesAsync(
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:758]: In the case where the coherency update fails, but there are no non-coherency updates, I think this method will erroneously calculate an empty PR description. I think it should probably have a reasonable description

**Review by mmitche — COMMENTED** [review]
Looking generally good. I haven't taken a very close look at the tests yet.

The only downside to the approach taken is that if a repo doesn't enable a merge policy, either using standard or including this specifically, they won't the direct feedback that there is an issue. IMO this isn't a blocker as standard is heavily used.

It would be good to eventually divide the merge policies into optional ones that enable auto-merge, and ones that are required to evaluate the dependency update.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 1 above)
Comment (mmitche) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:13]: Agreed.

Also this is just replacing `(repoDependencyUpdate.RequiredUpdates == null || !repoDependencyUpdate.RequiredUpdates,Any())`

IMO, that's clearer than `IsNullOrEmpty` against a collection anyway.

Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:13]: Updated. Using `.Count() < 1` instead.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 2 above)
Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:748]: >This looks like we'd create a PR if the coherency check currently fails? If so, this is great.

Yes. From [the comment](https://github.com/dotnet/arcade-services/pull/2266#issuecomment-1515177854): "We allow coherency to fail and still do an update PR. This is not the case today."

> Also, do we have a test for this corner case?

I have added a TestScenario `ScenarioTests_GitHubFlow.Darc_GitHubFlow_NonBatched_FailingStrictCoherency` that checks that the PR is opened while the coherency algorithm failed. And I've modified(+renamed) the existed test `PullRequestActorTests.UpdateWithAssetsWhenStrictAlgorithmFails`

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 3 above)
Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:758]: You are right, the commit description will be empty in this case. However, the GitHub check will contain all the required information. Additionally, when the customer resolves the coherency issue, the `UpdatePullRequestAsync` will update the description. I may add a description like ~ "Coherency update failed. Please refer to the GitHub checks for more details." But it sounds awkward for me. @riarenas and/or @mmitche what do you think,

File: src/Maestro/Maestro.MergePolicies/ValidateCoherencyMergePolicy.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+namespace Maestro.MergePolicies;
+
+public class ValidateCoherencyMergePolicy : MergePolicy
+{
+    public override string DisplayName => "Validate coherency";
+
+    public override Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemote darc) =>
+        Task.FromResult(pr.CoherencyCheckSuccessful ?
+            Succeed("Coherency check successful.") :
+            Fail("Coherency check failed.",
```
[the remark this answers is not in the document]
Comment (andriipatsula) [reply src/Maestro/Maestro.MergePolicies/ValidateCoherencyMergePolicy.cs]: Updated.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 2 above)
Comment (mmitche) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs]: The specific case here that I'm interested in is when there are no updates (except potential coherencyupdates), and coherency fails.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs (hunk 3 above)
Comment (mmitche) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:757]: I think we should title/describe the PR as if it was a coherency-only update like today, but note somewhere in the description that coherency has failed and the user should check the required checks,

If there are no commits because coherency failed, will AzDO and GitHub still open the PR? If not, maybe we add an empty commit with a commit message indicating that coherency failed.

Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:757]: Covered this use case:
* If coherent only update failed - we add an empty commit and description. Example: https://github.com/maestro-auth-test/maestro-test3/pull/2122
* Added additional scenario test to cover this use case

Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:757]: @mmitche could you please check the example and provide your feedback? Is the description and commit good enough?

Comment (andriipatsula) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:757]: Going to update: "Please check Github checks for details." --> "Please review the GitHub checks for more information."

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 5]
```diff
@@ -900,6 +906,15 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
                     coherencyUpdate.deps.Select(du => du.To).ToList(), message.ToString());
             }
 
+            // If the coherency algorithm failed and there are no non-coherency updates and
+            // we crate an empty commit that describes an issue.
+            if (requiredUpdates.Count == 0)
+            {
+                string message = "The Strict Coherency algorithm failed";
+                await remote.CommitUpdatesAsync(targetRepository, newBranchName, remoteFactory, new List<DependencyDetail>(), message);
+                return "This is a coherency update PR: " + message + ". Please review the GitHub checks for more information.";
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:915]:
```suggestion
                return $"This is a coherency update PR: " + message + ". Please review the GitHub checks for more information or run `darc update-dependencies --coherency-only` locally against {newBranchName}";
```

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs [hunk 6]
```diff
@@ -900,6 +906,15 @@ private async Task<string> CreatePullRequestAsync(List<UpdateAssetsParameters> u
                     coherencyUpdate.deps.Select(du => du.To).ToList(), message.ToString());
             }
 
+            // If the coherency algorithm failed and there are no non-coherency updates and
+            // we crate an empty commit that describes an issue.
+            if (requiredUpdates.Count == 0)
+            {
+                string message = "The Strict Coherency algorithm failed";
```
Comment (mmitche) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:913]:
```suggestion
                string message = "Failed coherent parent dependency update.";
```

I don't think we need to mention strict here, since strict becomes the only option.

File: src/Maestro/Maestro.MergePolicies/ValidateCoherencyMergePolicy.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    public override Task<MergePolicyEvaluationResult> EvaluateAsync(IPullRequest pr, IRemote darc)
+    {
+        if (pr.CoherencyCheckSuccessful)
+            return Task.FromResult(Succeed("Coherency check successful."));
+
+        StringBuilder description = new StringBuilder("Coherency update failed for the following dependencies:");
+        foreach (CoherencyErrorDetails error in pr.CoherencyErrors)
+        {
+            description.Append("\n * ").Append(error.Error);
```
Comment (riarenas) [inline src/Maestro/Maestro.MergePolicies/ValidateCoherencyMergePolicy.cs:27]: You can use AppendLine instead of Append so you can save yourself adding the \n manually

**Review by riarenas — APPROVED** [review]
LGTM pretty much. Just make sure to run all of the scenario test suite to make sure there are no problems there, and to keep in mind the necessary followups described in the issue.

**Review by riarenas — APPROVED** [review]
I did not look at the latest changes.

**Review by mmitche — APPROVED** [review]
